oop-sensor-data-analyzer/src/tests/CsvFileWriterTest.java

```diff
@@ -0,0 +1,168 @@
+            softly.assertThat(lines.get(41)).isEqualTo("CO2|Kitchen|0,10kg|08:00");
+            softly.assertThat(lines.get(42)).isEqualTo("CO2|Kitchen|0,30kg|18:00");
+            softly.assertThat(lines.get(43)).isEqualTo("CO2|Garage|0,20kg|08:00");
+            softly.assertThat(lines.get(44)).isEqualTo("CO2|Garage|0,15kg|18:00");
+            softly.assertThat(lines.get(45)).isEqualTo("CO2|Toilet|0,18kg|08:00");
+            softly.assertThat(lines.get(46)).isEqualTo("CO2|Toilet|0,32kg|18:00");
+            softly.assertThat(lines.get(47)).isEqualTo("TEMPERATURE|Garage|24,00°C|08:00");
+            softly.assertThat(lines.get(48)).isEqualTo("TEMPERATURE|Garage|30,00°C|18:00");
+            softly.assertThat(lines.get(49)).isEqualTo("TEMPERATURE|Inside|20,00°C|08:00");
+            softly.assertThat(lines.get(50)).isEqualTo("TEMPERATURE|Inside|25,00°C|18:00");
+            softly.assertThat(lines.get(51)).isEqualTo("TEMPERATURE|Outside|27,00°C|08:00");
+            softly.assertThat(lines.get(52)).isEqualTo("TEMPERATURE|Outside|35,00°C|18:00");
+            softly.assertThat(lines.get(53)).isEqualTo("HUMIDITY|Garage|34,00%|08:00");
+            softly.assertThat(lines.get(54)).isEqualTo("HUMIDITY|Garage|36,40%|18:00");
+            softly.assertThat(lines.get(55)).isEqualTo("HUMIDITY|Inside|30,00%|08:00");
+            softly.assertThat(lines.get(56)).isEqualTo("HUMIDITY|Inside|40,40%|18:00");
+            softly.assertThat(lines.get(57)).isEqualTo("HUMIDITY|Outside|50,00%|08:00");
+            softly.assertThat(lines.get(58)).isEqualTo("HUMIDITY|Outside|55,30%|18:00");
+            softly.assertThat(lines.get(59)).isEmpty();
+            softly.assertThat(lines.get(60)).isEqualTo("Highest latest readings :");
+            softly.assertThat(lines.get(61)).isEqualTo("CO2|Toilet|0,32kg|18:00");
+            softly.assertThat(lines.get(62)).isEqualTo("TEMPERATURE|Outside|35,00°C|18:00");
+            softly.assertThat(lines.get(63)).isEqualTo("HUMIDITY|Outside|55,30%|18:00");
+            softly.assertThat(lines.get(64)).isEmpty();
+            softly.assertThat(lines.get(65)).isEqualTo("Latest readings :");
+            softly.assertThat(lines.get(66)).isEqualTo("CO2|Kitchen|0,30kg|18:00");
+            softly.assertThat(lines.get(67)).isEqualTo("CO2|Garage|0,15kg|18:00");
+            softly.assertThat(lines.get(68)).isEqualTo("CO2|Toilet|0,32kg|18:00");
+            softly.assertThat(lines.get(69)).isEqualTo("TEMPERATURE|Garage|30,00°C|18:00");
+            softly.assertThat(lines.get(70)).isEqualTo("TEMPERATURE|Inside|25,00°C|18:00");
+            softly.assertThat(lines.get(71)).isEqualTo("TEMPERATURE|Outside|35,00°C|18:00");
+            softly.assertThat(lines.get(72)).isEqualTo("HUMIDITY|Garage|36,40%|18:00");
+            softly.assertThat(lines.get(73)).isEqualTo("HUMIDITY|Inside|40,40%|18:00");
+            softly.assertThat(lines.get(74)).isEqualTo("HUMIDITY|Outside|55,30%|18:00");
+        });
```

Comment:
your test looks cleaner and easier to understand - but it could be improved. now i am asking you to do an investigative task:
- What are the potential problems in this test?
- What could be improved?
Please note it is an open question and feel free to take it in a direction you want :) it's an association game and there are many good answers

Reply:
Maintainability is definitely an issue.
Whether it should be too strict or not depends on the use case, but I'd say in most cases trailing whitespaces don't matter, for example.
Parameterizing the test and applying `@CsvFileSource or @CsvFileSource`, for example, would add to the maintainability though - but for sure it's not an easy task. Please note that if we had parameterized this test case, we wouldn't have needed soft assertions, because parameterized test would have created different test cases for each one - but then you should have had to have extra private methods for setting up the test data itself. So yep.. It's tricky :P